policy.go

```diff
@@ -638,43 +682,102 @@ func readAndValidateLockNVIndexPublic(tpm *tpm2.TPMContext, index tpm2.ResourceC
 	}
 
 	pub.Attrs &^= tpm2.AttrNVReadLocked
 	// Validate its attributes
-	if pub.Attrs != lockNVIndexAttrs|tpm2.AttrNVWritten {
-		return nil, errors.New("unexpected NV index attributes")
+	if pub.Attrs != lockNVIndex1Attrs|tpm2.AttrNVWritten {
+		return nil, errors.New("unexpected index attributes")
 	}
 
 	clockBytes := make([]byte, binary.Size(clock))
 	binary.BigEndian.PutUint64(clockBytes, clock)
 
-	// Compute the expected authorization policy from the contents of the data index, and make sure this matches the public area.
+	// Compute the expected authorization policy from the contents of the data index, and make sure that this matches the public area.
 	// This verifies that the lock NV index has a valid authorization policy.
 	trial, err := tpm2.ComputeAuthPolicy(pub.NameAlg)
 	if err != nil {
-		return nil, xerrors.Errorf("cannot compute expected policy for NV index: %w", err)
+		return nil, xerrors.Errorf("cannot compute expected policy for index: %w", err)
 	}
 	trial.PolicyCommandCode(tpm2.CommandNVWrite)
 	trial.PolicyCounterTimer(clockBytes, 8, tpm2.OpUnsignedLT)
 	trial.PolicySigned(keyName, nil)
 
 	if !bytes.Equal(trial.GetDigest(), pub.AuthPolicy) {
-		return nil, errors.New("incorrect policy for NV index")
+		return nil, errors.New("incorrect policy for index")
 	}
 
-	// This is a valid global lock NV index that cannot be recreated!
+	// pub corresponds to a valid legacy global lock NV index that cannot be recreated!
 	return pub, nil
 }
 
+// validateCurrentLockNVIndices validates that the supplied NV indices correspond to valid new-style lock NV indices.
+func validateCurrentLockNVIndices(tpm *tpm2.TPMContext, index1, index2 tpm2.ResourceContext, session tpm2.SessionContext) error {
+	var s tpm2.SessionContext
+	if session != nil {
+		s = session.IncludeAttrs(tpm2.AttrAudit)
+	}
+
+	index1Pub, _, err := tpm.NVReadPublic(index1, s)
+	if err != nil {
+		return xerrors.Errorf("cannot read public area of index 1: %w", err)
+	}
+	index1Pub.Attrs &^= tpm2.AttrNVReadLocked
+	index1Name, err := index1Pub.Name()
+	if err != nil {
+		return xerrors.Errorf("cannot compute name of index 1: %w", err)
+	}
+
+	index1ExpectedName, index2ExpectedName, err := computeLockNVIndexNames()
+	if err != nil {
+		return xerrors.Errorf("cannot compute expected names: %w", err)
+	}
+
+	if !bytes.Equal(index1Name, index1ExpectedName) || !bytes.Equal(index2.Name(), index2ExpectedName) {
+		return errors.New("found indices with unexpected names")
+	}
+
+	return nil
+}
+
+// validateLockNVIndices checks that the NV indices at the global handles used for locking access to sealed keys are valid lock
+// indices, and returns an error if they aren't. If the indices correspond to the legacy mechanism, the public area of the lock index
+// is returned, else no public area is returned.
+func validateLockNVIndices(tpm *tpm2.TPMContext, session tpm2.SessionContext) (*tpm2.NVPublic, error) {
```

Comment:
I understand it somehow minimized the diff, because OTOH the code base is not very internally modular ATM, but it's very awkward as a signature to carry into the future that this returns the legacy lock index public area or not.

I have an idea for an immediate follow-up, that would solve maybe that and also explore patterns to make the code a bit more internally modular.

Reply:
@chrisccoulson I proposed the follow-up  https://github.com/snapcore/secboot/pull/118, it's in draft because some things are best moved once this has conflict fixed and the doc comment for ensureLockNVIndices updated. See last commit there: 8c62bf263ed3b451